src/network/taraxa_capability.cpp

```diff
@@ -400,21 +425,36 @@ void TaraxaCapability::interpretCapabilityPacketImpl(NodeID const &_nodeID, unsi
       break;
     }
     case GetBlocksPacket: {
-      LOG(log_dg_dag_sync_) << "Received GetBlocksPacket with " << _r.itemCount() << " known blocks";
-      std::unordered_set<blk_hash_t> known_non_finalized_blocks;
-      for (auto hash : _r) {
-        known_non_finalized_blocks.insert(blk_hash_t(hash));
-      }
+      std::unordered_set<blk_hash_t> blocks_hashes;
       std::vector<std::shared_ptr<DagBlock>> dag_blocks;
+      auto mode = static_cast<GetBlocksPacketRequestType>(_r[0].toInt<unsigned>());
+
+      if (mode == MissingHashes)
+        LOG(log_dg_dag_sync_) << "Received GetBlocksPacket with " << _r.itemCount() - 1 << " missing blocks";
+      else
+        LOG(log_dg_dag_sync_) << "Received GetBlocksPacket with " << _r.itemCount() - 1 << " known blocks";
+
+      for (size_t i = 1; i < _r.itemCount(); ++i) {
+        blocks_hashes.emplace(_r[i]);
+      }
+
       const auto &blocks = dag_mgr_->getNonFinalizedBlocks();
       for (auto &level_blocks : blocks) {
         for (auto &block : level_blocks.second) {
           auto hash = blk_hash_t(block);
-          if (known_non_finalized_blocks.count(hash) == 0) {
-            dag_blocks.emplace_back(db_->getDagBlock(hash));
+          if (mode == MissingHashes) {
+            if (blocks_hashes.count(hash) == 1) dag_blocks.emplace_back(db_->getDagBlock(hash));
+          } else {
+            if (blocks_hashes.count(hash) == 0) dag_blocks.emplace_back(db_->getDagBlock(hash));
           }
         }
       }
+
+      // This means that someone requested more hashes that we actually have -> do not send anything
+      if (mode == MissingHashes && dag_blocks.size() != blocks_hashes.size()) {
+        LOG(log_nf_dag_sync_) << "Node " << _nodeID << " requested uknow DAG block from us";
```

Comment:
- maybe at least warning ? 
- typo in "uknow"
- it's not "us" I guess  :D and again last 2 words useless info without any additional value

Reply:
I would not mark it as warning as this is not issue of this node.